### ISSUE TITLE
signed-by is unusable in sources.list

### ISSUE THREAD

**Issue by Cornelicorn** [systemd/mkosi#2065]
When using a sources.list entry with `signed-by=/etc/apt/keyrings/something.asc`, apt will look for this key on the host, not inside the image.
This can be circumvented by installing the sources.list entry in a package manager tree with an adjusted path to the file in your config and with the "correct" path inside an extra tree instead of using a skeleton tree, but this is inconvenient.

I'll have to look into the apt source code whether this is something mkosi can solve on it's on or whether this needs changes inside apt with an additional option for the root directory to search these key files in added.

**Comment by Cornelicorn** [comment]
> I'll have to look into the apt source code whether this is something mkosi can solve on it's on or whether this needs changes inside apt with an additional option for the root directory to search these key files in added.

Doesn't seem to be the case, the relevant functions are `GPGVMethod::VerifyGetSigners` used by `GPGVMethod::VerifyGetSignersWithLegacy`, which calls `ExecGPGV`, which just appends the value found in the `source.list` entry as a `--keyring` argument to `apt-key`. The binary used for apt-key can be overriden though. So a dirty hack might be to install a binary which modifies each --keyring argument to prepend the working dir if the argument is an absolute path and is not relative to the working dir already.

More elaborate: Create mkosi-apt-key, which when called with `--keyring /etc/apt/keyrings/something.asc ...` runs `apt-key --keyring /.../mkosi.XXXXX/etc/apt/keyrings/something.asc ...`.
The relevant setting is `Dir::Bin::apt-key` to set the apt-key binary used.

**Comment by Cornelicorn** [comment]
As a first draft, I created this script, called it `apt-key-mkosi`
```
#!/bin/bash

declare -a args
declare prefix

prefix="${TMPDIR}/pkgmngr/"

while [[ "$#" -gt 0 ]]
do
    case $1 in
        --keyring)
            args+=("--keyring")
            inside_location="${prefix}/${2}"
            if [ -f "$inside_location" ]
            then
                args+=("$inside_location")
            else
                args+=("$2")
            fi
            shift 2
        ;;
        *)
            args+=("$1")
            shift
        ;;
    esac
done

command apt-key "${args[@]}"
```
and set `Dir::Bin::apt-key /path/to/apt-key-mkosi;` inside `mkosi.skeleton/etc/apt/apt.conf.d/30aptkey`.
This let's you use keyrings inside the package manager trees and skeleton trees and they would get preffered over those in the host image.
I think this would be nice to have this script (or an equivalent implementation in python) shipped within mkosi, s.t. apt-key calls are automatically adjusted. 
I'm not sure whether the `TMPDIR` variable used here is considered stable?

**Comment by behrmann** [comment]
I think something like this would be useful, though I'm not sure whether overlaying the package manager tree over a tools tree might not be clearer, but it would probably need both, since we can't guarantee the use of a tools tree.

>  I'm not sure whether the `TMPDIR` variable used here is considered stable?

We don't list it in the environment variable table on the other hand we always set it. Nevertheless, I think the point is moot, since this would be for mkosi internal use.

**Comment by Cornelicorn** [comment]
> We don't list it in the environment variable table on the other hand we always set it. Nevertheless, I think the point is moot, since this would be for mkosi internal use.

Yes this was more in the direction of "If this is not wanted directly in mkosi, is there an alternative to using TMPDIR which will be usable in the foreseeable future", since using the host keyrings was openly communicated in `NEWS`.

**Comment by DaanDeMeyer** [comment]
Fixed by #2210
